### PR TITLE
Import native binlog files

### PR DESCRIPTION
**Pull request by yhmo**
Signed-off-by: yhmo <yihua.mo@zilliz.com>

#15604

### CODE REVIEW

File: internal/util/importutil/import_wrapper.go [hunk 1]
```diff
@@ -27,7 +44,21 @@ import (
...
+
+	// this limitation is to avoid this OOM risk:
+	// simetimes system segment max size is a large number, a single segment fields data might cause OOM.
+	// flush the segment when its data reach this limitation, let the compaction to compact it later.
+	MaxSegmentSizeInMemory = 512 * 1024 * 1024 // 512MB
+
+	// this limitation is to avoid this OOM risk:
+	// if the shard number is a large number, although single segment size is small, but there are lot of in-memory segments,
+	// the total memory size might cause OOM.
+	MaxTotalSizeInMemory = 2 * 1024 * 1024 * 1024 // 2GB
```
Comment (soothing-rain) [inline internal/util/importutil/import_wrapper.go:61]: IIUC, maxTotalSize should be equal to the file size limit on a Bulk Load task, which is 1GigBytes.

Why are we setting it 2GigBytes here while loading backup files?

Comment (yhmo) [reply internal/util/importutil/import_wrapper.go:61]: No, for column-based task, there are several files, each file limit is 1GB, but the total size may exceed 1GB

**Comment by yhmo** [comment]
rerun ut

File: internal/util/importutil/import_wrapper.go [hunk 2]
```diff
@@ -156,6 +187,14 @@ func (p *ImportWrapper) fileValidation(filePaths []string, rowBased bool) error
 // filePath and rowBased are from ImportTask
 // if onlyValidate is true, this process only do validation, no data generated, callFlushFunc will not be called
 func (p *ImportWrapper) Import(filePaths []string, rowBased bool, onlyValidate bool) error {
+	log.Info("ImportWrapper: filePaths", zap.Any("filePaths", filePaths))
+	// data restore function to import milvus native binlog files(for backup/restore tools)
+	// the backup/restore tool provide two paths for a collection, the first path is binlog path, the second is deltalog path
+	if p.isBinlogImport(filePaths) {
```
Comment (soothing-rain) [inline internal/util/importutil/import_wrapper.go:193]: So we are not calling `debug.FreeOSMemory()` in while importing binlogs? Could it be a problem when we have a lot of binlog files?

File: internal/util/importutil/import_wrapper.go [hunk 3]
```diff
@@ -254,20 +293,68 @@ func (p *ImportWrapper) Import(filePaths []string, rowBased bool, onlyValidate b
 		}
 	}
 
+	return p.reportPersisted()
```
Comment (soothing-rain) [inline internal/util/importutil/import_wrapper.go:296]: `debug.FreeOSMemory()` is only called once, when all data has been recovered/bulk-loaded.

Would it be a problem when there's a lot of data? Maybe we should call `debug.FreeOSMemory()` for every `MaxTotalSizeInMemory` size of data?

**Comment by yhmo** [comment]
rerun ut

File: internal/util/importutil/binlog_adapter.go [hunk 4]
```diff
@@ -0,0 +1,1083 @@
...
+		}
+	}
+
+	return nil
+}
+
+// This method read data from deltalog, and convert to a dict
+// The deltalog data is a list, to improve performance of next step, we convert it to a dict,
+// key is the deleted ID, value is operation timestamp which is used to apply or skip the delete operation.
+func (p *BinlogAdapter) readDeltalogs(segmentHolder *SegmentFilesHolder) (map[int64]uint64, map[string]uint64, error) {
```
[the remark this answers is not in the document]
Comment (yhmo) [reply internal/util/importutil/binlog_adapter.go:290]: Wrap the map[int64]uint64, map[string]uint64 to a struct like this?
```
type IDArray struct {
  map[int64]uint64 
  map[string]uint64
}

readDeltalogs(segmentHolder *SegmentFilesHolder) (*IDArray, error)
```

File: internal/util/importutil/binlog_adapter.go [hunk 5]
```diff
@@ -0,0 +1,1083 @@
...
+	// primary key not found
+	if adapter.primaryKey == -1 {
+		log.Error("Binlog adapter: collection schema has no primary key")
+		return nil, errors.New("collection schema has no primary key")
+	}
+
+	return adapter, nil
+}
+
+func (p *BinlogAdapter) Read(segmentHolder *SegmentFilesHolder) error {
```
[the remark this answers is not in the document]
Comment (yhmo) [reply internal/util/importutil/binlog_adapter.go:119]: This work ought to be done in a future pr with the following steps:
1. refine the logic of reading deltalog and insertlog, define common interfaces
2. implement common interfaces in storage package
3. datanode/compactor.go and util/importuitl/binlog_adapter.go use same interfaces of storage package

But this pr shouldn't include this work.

File: internal/util/importutil/import_wrapper.go (hunk 3 above)
Comment (yhmo) [reply internal/util/importutil/import_wrapper.go:321]: debug.FreeOSMemory() returning all free memory back to the underlying system at once, this operation is expensive and can lead to latency spikes as it holds the heap lock through the whole process, especially in standalone mode.

Comment (yhmo) [reply internal/util/importutil/import_wrapper.go:335]: I have changed the logic to call debug.FreeOSMemory() after each segment imported.

**Comment by xiaofan-luan** [comment]
I will just give a weak yes.

For the reason of backup restore test, I will approve this issue. but definitely need more refactor on this.

**Pull request merged**